### DISCUSSION TITLE
Hook auth.login.before

### DISCUSSION THREAD

**Discussion by NilsBaumgartner1994** [directus/directus#9651] [Archive]
In the old version i could receive the profile from the profile url of an 0auth service. now there is a payload with code and codeVerifier which doesnt help much. How can i get the profile data ?

**Comment by joselcvarela** [comment]
Hello @NilsBaumgartner1994 
Have you adapted your custom hooks to match the new structure?
https://github.com/directus/directus/blob/main/changelog.md#%EF%B8%8F-breaking-changes
https://docs.directus.io/extensions/hooks/#custom-api-hooks

**Reply by NilsBaumgartner1994** [reply]
Yes, but unfortunately is only gives me now a payload instead of the profile.

`module.exports = function registerHook({ services, exceptions, database, getSchema, env, logger }) {

  return {
    "auth.login.before": async function (input) {
    const payload = input.payload;
    console.log(input);
    console.log(payload);
    ....
`


This will give me the folling output:

`
myappdirectus |   event: 'auth.login.before',
myappdirectus |   action: 'login',
myappdirectus |   schema: {....
myappdirectus |   },
myappdirectus |   payload: {
myappdirectus |     code: 'fa98a89480bec2608d76df17de9c8910d1484074',
myappdirectus |     codeVerifier: 'Tg0NpphxCcFshaEOUbAEG5Z0MfgyBrM8TiMCJKmpKYE'
myappdirectus |   },
myappdirectus |   provider: 'customOauth',
myappdirectus |   accountability: {
myappdirectus |     ip: '172.19.0.1',
myappdirectus |     userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/95.0.4638.69 Safari/537.36',
myappdirectus |     role: null
myappdirectus |   },
....
`


So the payload itself:
`payload: {
     code: 'fa98a89480bec2608d76df17de9c8910d1484074',
     codeVerifier: 'Tg0NpphxCcFshaEOUbAEG5Z0MfgyBrM8TiMCJKmpKYE'
}`

does not give me much information ...

**Reply by NilsBaumgartner1994** [reply]
Sorry for bad formating

**Reply by joselcvarela** [reply]
This seems the old version.
Currently you will need to do something like

```js
filter('auth.login', async (payload) => ...
// OR
action('auth.login', ({payload}) => {
```

On those you should have the payload sent by the client.

**Reply by NilsBaumgartner1994** [reply]
Oh yes sorry, was dealing with version 100, which was before some days the newest.

**Reply by NilsBaumgartner1994** [reply]
Nice, now none of the hooks are getting called...

15:51:40 ✨ Loaded extensions: createSSOUsers

But none of the hooks are fired:

`module.exports = function registerHook({ filter, action, init, schedule }, { services, exceptions, database, getSchema, env, logger }) {
  const { ServiceUnavailableException, ForbiddenException } = exceptions;

  action('auth.login.before', (input) => {
    throw new ForbiddenException();
    console.log("extensions/hooks/createSSOUsers - action('auth.login.before'");
    logger.info("extensions/hooks/createSSOUsers - action('auth.login.before'")
    console.log(input);
    return input;
  });

  filter('auth.login.before', (input) => {
    throw new ForbiddenException();
    console.log("extensions/hooks/createSSOUsers - filter('auth.login.before'");
    logger.info("extensions/hooks/createSSOUsers - filter('auth.login.before'")
    console.log(input);
    return input;
  });

};`

**Reply by NilsBaumgartner1994** [reply]
Logins should now be imposible, but i still manage to get in and the console shows no information...

**Reply by NilsBaumgartner1994** [reply]
Okay problem was, hat not ".before" does not exist.

**Reply by NilsBaumgartner1994** [reply]
Getting the event but again, the payload/input is not nice...

myappdirectus | {
myappdirectus |   payload: {
myappdirectus |     code: '27c2fb6e2b1cd23cea1e32447d3af85e9e8abfe2',
myappdirectus |     codeVerifier: 'j_TXCES2uynAEsiRJ6cisGzxLEayzhCvlSBG28UsSkM',
myappdirectus |     state: 'DbmT1J0uYRiugGpFUMal0d8a1TZKQj_lC7vg21vxp2k'
myappdirectus |   },
myappdirectus |   status: 'success',
myappdirectus |   user: '55db28d1-df60-4769-bc0c-04ced2ee87bb',
myappdirectus |   provider: 'adobe'
myappdirectus | }

**Comment by NilsBaumgartner1994** [comment]
Done it the hard "hacky" way.
Put this code inside your extension. Works for oAuth2 clients.

const openIDClient = require('openid-client');

const getUserProfile = async (userId, provider, database, env) => {
  const existingUser = await database("directus_users")
      .where({ id: userId })
      .first();
  let providerUppercase = provider.toUpperCase();

  const  { Issuer } = openIDClient;

  let auth_data = existingUser.auth_data;
  let tokenSet = auth_data;

  let envPre = "AUTH_";
  let clientId = env[envPre+providerUppercase+"_CLIENT_ID"];
  let clientSecret = env[envPre+providerUppercase+"_CLIENT_SECRET"];
  let authorizeUrl = env[envPre+providerUppercase+"_AUTHORIZE_URL"];
  let accessUrl = env[envPre+providerUppercase+"_ACCESS_URL"];
  let profileUrl = env[envPre+providerUppercase+"_PROFILE_URL"];
  let additionalConfig = {provider: provider};

  console.log("create now issuer")
  let issuer = new Issuer({
    authorization_endpoint: authorizeUrl,
    token_endpoint: accessUrl,
    userinfo_endpoint: profileUrl,
    issuer: additionalConfig.provider,
  });

  console.log("create now redirectURL")
  let urlSegments = [env.PUBLIC_URL,'auth','login', additionalConfig.provider, 'callback']
  let redirectUrl = urlSegments.join('/');

  console.log("create now client")
  let client = new issuer.Client({
    client_id: clientId,
    client_secret: clientSecret,
    redirect_uris: [redirectUrl],
    response_types: ['code'],
  });

  console.log("get now userInfo");
  let userInfo = await client.userinfo(tokenSet);
  console.log(userInfo);
  
  return userInfo;
}

**Reply by NilsBaumgartner1994** [reply]
A nicer updated gist

https://gist.github.com/NilsBaumgartner1994/be374e0aef5d92012451e84d36677dac
